---
title: "The pipelie for identification of the differentially-interacted regions (DIRs)."
author: "Tsunghan Hsieh"
description: > This notebook demonstrates how I get DIRs from human samples treated with microbial molecules.
date: "`r format(Sys.Date(), format='%d-%m-%Y')`"
output:
 epuRate::epurate:
 toc: TRUE
 number_sections: TRUE
 code_folding: "show"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set up
***
```{r}
# Load functions.
devtools::document()
devtools::load_all()
```

# Library
***
```{r library}
library(edgeR)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(GenomicRanges)
library(GenomicFeatures)
library(AnnotationDbi)
```

# Set path
***

```{r path}
inputDir <- here::here("./Results/processing")
outputDir <- here::here("./Results/DIR")
```

# TODO
***
```{python todo}
# Before running this document, please run the following python script to subset hic data 
# and generate the position index file
# Example
# python ./Python/subset_hic_data.py --inputDir /data/44111_A_ctrl_43614_mc5contact_map.hic --outputDir /Results/processing/ --rowChr chr2 --columnChr chr2 --rowStart 112735986 --rowEnd 113204585 --columnStart 112735986 --columnEnd 113204585 --res 5000
```

# Read the input file
*** 

```{r input}
hicfiles <- list.files(path = inputDir, pattern = "_map_extracted\\.csv$", full.names = TRUE)
posfiles <- list.files(path = inputDir, pattern = "_map_PosIndex\\.txt$", full.names = TRUE)
```

# Generate the DIR
***
We will generate a series of DIR files. The first one is the differential analysis among cells from two healthy donors and this file will be used as a "black list" to filter out the individual-specific interacted regions. Other DIR files are human cells treated with microbial molecules vs untreated. Note we have no biological replicate for all experiment, therefore we assume biological covariance = 0.4 for the differential analysis.

## DIR dataframe
***

```{r DIR}
## Healthy donors
df_healthy <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/44111_A_ctrl_43614_mc5contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/44113_C_ctrl_43616_mc7contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/44111_A_ctrl_43614_mc5contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/healthy.csv")
)

## Stimulation
df_bg <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/44112_A_bg_43615_mc6contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/44111_A_ctrl_43614_mc5contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/44111_A_ctrl_43614_mc5contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bg.csv")
)

df_lps <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/44114_C_lps_43617_mc8contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/44113_C_ctrl_43616_mc7contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/44113_C_ctrl_43616_mc7contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/lps.csv")
)

df_bcg <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/BCG_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/DMSO_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bcg.csv")
)

df_ldhi <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/LDHi_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/DMSO_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/ldhi.csv")
)

df_bcgldhi <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/BCGLDHi_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/DMSO_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bcgldhi.csv")
)

## DIR across different stimulation conditions
## Xiaolin exp
df_bcgldhi_bcg <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/BCGLDHi_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/BCG_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bcgldhi_bcg.csv")
)

df_bcgldhi_ldhi <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/BCGLDHi_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/LDHi_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bcgldhi_ldhi.csv")
)

df_bcg_ldhi <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/BCG_contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/LDHi_contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/DMSO_contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bcg_ldhi.csv")
)

## Me and Winsen
df_bg_lps <- GetDIRWithNoReplicate(
chr = "chr2",
treat = here::here("./Results/processing/44112_A_bg_43615_mc6contact_map_extracted.csv"),
ctrl = here::here("./Results/processing/44114_C_lps_43617_mc8contact_map_extracted.csv"),
bcv = 0.4,
start_position_index = here::here("./Results/processing/44111_A_ctrl_43614_mc5contact_map_PosIndex.txt"),
output = here::here("./Results/DIR/bg_lps.csv")
)

```
## Annotate the DIR
***
```{r annotation}
annoted_df_list <- list()
for (treat in c("healthy", "bg", "lps", "bcg", "ldhi", "bcgldhi",
                "bcgldhi_bcg", "bcgldhi_ldhi", "bcg_ldhi", "bg_lps")) {
  input = here::here(paste0("./Results/DIR/", treat, ".csv"))
  output = here::here(paste0("./Results/DIR/annotated_", treat, ".csv"))
  annoted_df_list[[treat]] <- annotateDIR(
    input = input,
    output = output
  )
}
```

# (Optional) Filter out the individual-specific DIRs
***

```{r filter_blacklist}
## looks like only a small fraction are individual-specific
df_blacklist <- annoted_df_list[["healthy"]] |>
  dplyr::filter(PValue < 0.1) |>
  dplyr::mutate(ixid = paste0(region1, "-", region2))

write.csv(df_blacklist, here::here("./Results/DIR/annotated_blacklist.csv"))

filtered_annoted_df_list <- list()
for (treat in c("healthy", "bg", "lps", "bcg", "ldhi", "bcgldhi",
                "bcgldhi_bcg", "bcgldhi_ldhi", "bcg_ldhi", "bg_lps")) {
  filtered_annoted_df_list[[treat]] <- annoted_df_list[[treat]] |>
    dplyr::mutate(ixid = paste0(region1, "-", region2)) |>
    dplyr::filter(!ixid %in% df_blacklist$ixid)
  write.csv(filtered_annoted_df_list[[treat]],
            here::here(paste0("./Results/DIR/filtered_annotated_", treat, ".csv")))
}

```

# Identify DIRs
***
```{r dir_summary}
DIR_df <- list()
for (treat in c("healthy", "bg", "lps", "bcg", "ldhi", "bcgldhi",
                "bcgldhi_bcg", "bcgldhi_ldhi", "bcg_ldhi", "bg_lps")) {
  DIR_df[[treat]] <- annoted_df_list[[treat]] |>
    dplyr::filter(PValue < 0.05)
  write.csv(DIR_df[[treat]],
            here::here(paste0("./Results/DIR/annotated_DIR_", treat, ".csv")))
}
```

# Identify the interaction between IL-1B and IL-37
***
Musa is interested in the interaction between IL-1B and IL-37. Therefore we filtered out DIRs from these two regions. We will also arrange the dataframe into bed files, so that we can upload it to UCSC genome browser.

Here are the columns for the BED file format. Note there should be no header in the BED file.

1. chrom - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).
2. chromStart - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered 0.
3. chromEnd - The ending position of the feature in the chromosome or scaffold. 

The 9 additional optional BED fields are:

4. name - Defines the name of the BED line. This label is displayed to the left of the BED line in the Genome Browser window when the track is open to full display mode or directly to the left of the item in pack mode.
5. score - A score between 0 and 1000. If the track line useScore attribute is set to 1 for this annotation data set, the score value will determine the level of gray in which this feature is displayed (higher numbers = darker gray). This table shows the Genome Browser's translation of BED score values into shades of gray.
6. strand - Defines the strand. Either "." (=no strand) or "+" or "-".

7 - 10 are not relevant to this project.

```{r il1b-il37}
# IL1b: chr2:112,829,751-112,836,779 
# IL37: chr2:112,911,165-112,918,882

# The data is diagonal symmetric, there we just need to check one gene in each region
resolution <- 5000
il1b_upstream <- 112829751 - 5000
il1b_downstream <- 112836779 + 5000 
il37_upstream <- 112911165 - 5000
il37_downstream <- 112918882 + 5000
filtered_annotated_il1bil37_df_list <- list()

## filtered out by black list does not chane the result
for (treat in c("healthy", "bg", "lps", "bcg", "ldhi", "bcgldhi",
                "bcgldhi_bcg", "bcgldhi_ldhi", "bcg_ldhi", "bg_lps")) {
  filtered_annotated_il1bil37_df_list[[treat]] <- filtered_annoted_df_list[[treat]] |>
    dplyr::filter((region1 - resolution/2) > il1b_upstream & (region1 + resolution/2) < il1b_downstream) |>
    dplyr::filter((region2 - resolution/2) > il37_upstream & (region2 + resolution/2) < il37_downstream)
  write.csv( filtered_annotated_il1bil37_df_list[[treat]],
            here::here(paste0("./Results/DIR/filtered_annotated_il1bil37_", treat, ".csv")))
}


## Arrange the dataframe to bed format
filtered_annotated_il1bil37_bed_list <- list()
for (treat in c("healthy", "bg", "lps", "bcg", "ldhi", "bcgldhi",
                "bcgldhi_bcg", "bcgldhi_ldhi", "bcg_ldhi", "bg_lps")) {
  df <- read.csv(
    here::here(paste0("./Results/DIR/filtered_annotated_il1bil37_", treat, ".csv")))
  df <- df[,2:ncol(df)]
  final_df <- df[,1:3]
  final_df$treat <- treat
  final_df <- cbind(final_df, data.frame(logFC = df[["logFC"]]))
  final_df$strand <- "+"
  filtered_annotated_il1bil37_bed_list[[treat]] <- final_df
  write.table(filtered_annotated_il1bil37_bed_list[[treat]],
             here::here(paste0("./Results/DIR/filtered_annotated_il1bil37_", treat, ".bed")), sep = "\t", col.names = FALSE, row.names = FALSE, quote = FALSE)
}
remove(df)

```