#' annotateDIR
#'
#' @description
#' This function reads differentially-interacted regions (DIRs) and annotate with hg38 information
#' 
#' @author Tsunghan Hsieh
#'
#' @param input a character string specifying the name and path of the input file (.csv)
#' This file should be generated by getDIRWithNoReplicate()
#' @param output A character string specifying the name and path of output file
#' the default path is as same as the treatment file
#' the default output file name is the treatment file name prefixed with "annotated_"
#' @return a dataframe of the output data
#'
#' @export
#'
#' @importFrom here here
#' @importFrom assertthat assert_that
#' @importFrom tidyr separate
#' @importFrom TxDb.Hsapiens.UCSC.hg38.knownGene
#' @importFrom org.Hs.eg.db
#' @importFrom GenomicRanges
#' @importFrom AnnotationDbi select
#'
#' @examples
#' \dontrun{
#' df <- annotateDIR(
#' input = here::here("./Results/DIR/DIR_44112_A_bg_43615_mc6contact_map_extracted.csv"),
#' output = NULL
#' )
#' }
#' 
## TODO
## Run getDIRWithNoReplicate() to generate DIR files

annotateDIR <- function(input, output) {
  # check input ----------------------------------------------------------------
  
  assertthat::assert_that(is.character(input), 
                          msg = "The given argument is not a string.\n")
  
  assertthat::assert_that(grepl("\\.csv$", input), 
                          msg = "The given file to the input file is not a .csv file.\n")
  
  assertthat::assert_that(file.exists(here::here(input)), 
                          msg = "The given file path to the input file does not exist.\n")
  
  # check output ---------------------------------------------------------------
  
  if (is.null(output)) {
    # Use sub() to extract the desired part
    outputfname <- sub(".*/([^.]+)\\.csv", "\\1", input)
    outputfname <- paste0("annotated_", outputfname, ".csv")
    outputdir <- dirname(input)
    output <- here::here(outputdir, outputfname)
    remove(outputfname, outputdir)
  } else {
    assertthat::assert_that(is.character(output), 
                            msg = "The given argument is not a string.\n")
    
    assertthat::assert_that(grepl("\\.csv$", output), 
                            msg = "The given file to the output file is not a .csv file.\n")
    
    assertthat::assert_that(file.exists(here::here(dirname(output))), 
                            msg = "The given file path to the expertiment file does not exist.\n")
  }
  
  
  # Assign the hg38 objects ----------------------------------------------------
  .GlobalEnv$txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
  hgdb <- org.Hs.eg.db
  
  # Extract TSS positions ------------------------------------------------------
  tss <- transcripts(txdb, columns=c("tx_id", "gene_id"), use.names=TRUE)
  tss <- resize(tss, width=1, fix='start')
  tss_df <- as.data.frame(tss, row.names = NULL)
  tss_df$transcript <- row.names(tss_df)
  colnames(tss_df) <- c("tx_chr","tx_start","tx_end","tx_width","tx_strand","tx_id","gene_id","transcript")
  
  # Read DIRs ------------------------------------------------------------------
  input_df <- read.csv(input)
  input_df <- input_df[,2:ncol(input_df)]
  
  annotated_df <- .get_annotated_gr(input_df)
  write.csv(annotated_df, output)
  message("Done! The output is written in: ", output)
}


#' .get_annotated_gr
#'
#' @description
#' This function annotate differentially-interacted regions (DIRs) with given genome information
#' 
#' @author Tsunghan Hsieh
#'
#' @param input a character string specifying the input dataframe
#' @return a dataframe of annotated file
#' @examples
#' \dontrun{
#' df <- .get_annotated_gr(
#' df = input_df
#' )
#' }
.get_annotated_gr <- function(df) {
  
  # Check colnames -------------------------------------------------------------
  essential_col <- c("chr", "region1", "region2")
  for (i in 1:length(essential_col)) {
    if (!essential_col[i] %in% colnames(df)) {
      message("Please make sure you have the following column: ", essential_col[i])
      return (0)
    } 
  }
  remove(essential_col, i)
  
  ## Filter out non-standard chr
  wanted <- paste0("chr", seq(1:23))
  wanted <- c(wanted, "chrX", "chrY")
  df <- df |>
    dplyr::filter(chr %in% wanted) |>
    dplyr::mutate(region1 = as.numeric(region1),
                  region2 = as.numeric(region2)) |>
    dplyr::mutate(ix_id = paste0(chr, "_", region1, "_", region2))
  
  # Annotate the region1 and region2 seperately --------------------------------
  ## For creating GRange objects
  ## The ix_id column is used as the identity column
  
  df1 <- df |>
    dplyr::select(chr, region1, ix_id) |>
    dplyr::mutate(start = region1,
                  end = region1+1)
  
  df2 <- df |>
    dplyr::select(chr, region2, ix_id) |>
    dplyr::mutate(start = region2,
                  end = region2+1)

  
  # Create GRange objects ------------------------------------------------------
  gr1 <- GRanges(
    seqnames = df1$chr,
    ranges = IRanges(start = df1$start, end = df1$end),
    ix_id = df1$ix_id
  )
  
  gr2 <- GRanges(
    seqnames = df2$chr,
    ranges = IRanges(start = df2$start, end = df2$end),
    ix_id = df2$ix_id
  )
  
  # Finding overlap ------------------------------------------------------------
  features <- transcripts(txdb)
  
  # Select the id of closet genes ----------------------------------------------
  ## nearest from peak start
  ps.idx1 <- follow(gr1, features)
  ps.idx2 <- follow(gr2, features)
  
  ## nearest from peak end
  pe.idx1 <- precede(gr1, features)
  pe.idx2 <- precede(gr2, features)
  
  ## check if any peak not matched to the nearest gene
  na.idx1 <- is.na(ps.idx1) & is.na(pe.idx1)
  na.idx2 <- is.na(ps.idx2) & is.na(pe.idx2)
  
  ## Remove na 
  if (sum(na.idx1) > 0) { ## suggested by Thomas Schwarzl
    ps.idx1 <- ps.idx1[!na.idx1]
    pe.idx1 <- pe.idx1[!na.idx1]
    gr1 <- gr1[!na.idx1]
  }
  
  if (sum(na.idx2) > 0) { ## suggested by Thomas Schwarzl
    ps.idx2 <- ps.idx2[!na.idx2]
    pe.idx2 <- pe.idx2[!na.idx2]
    gr2 <- gr2[!na.idx2]
  }
  
  # Generate the final data frame ----------------------------------------------
  ## Identify the closet gene information
  features_df1 <- data.frame(txStart1 = start(features[ps.idx1]),
                            txEnd1 = end(features[pe.idx1]),
                            txStrand1 = strand(features[ps.idx1]),
                            tx_id1 = features[ps.idx1]$tx_id,
                            tx_name1 = features[ps.idx1]$tx_name)
  
  features_df2 <- data.frame(txStart2 = start(features[ps.idx2]),
                             txEnd2 = end(features[pe.idx2]),
                             txStrand2 = strand(features[ps.idx2]),
                             tx_id2 = features[ps.idx2]$tx_id,
                             tx_name2 = features[ps.idx2]$tx_name)
  
  features_df1 <- features_df1 |> 
    dplyr::mutate(distanceToTSS1 = ifelse(txStrand1 == "+", txEnd1 - txStart1, txStart1 - txEnd1))
  
  features_df2 <- features_df2 |> 
    dplyr::mutate(distanceToTSS2 = ifelse(txStrand2 == "+", txEnd2 - txStart2, txStart2 - txEnd2))
  
  gr_df1 <- as.data.frame(gr1)
  annotated_df1 <- cbind(gr_df1, features_df1)
  annotated_df1 <- annotated_df1[,6:ncol(annotated_df1)]
  
  gr_df2 <- as.data.frame(gr2)
  annotated_df2 <- cbind(gr_df2, features_df2)
  annotated_df2 <- annotated_df2[,6:ncol(annotated_df2)]
  
  
  ## map to the gene ids
  mapped_genes1 <- select(txdb, 
                         keys = annotated_df1$tx_name, 
                         keytype = "TXNAME", 
                         columns = c("GENEID"))
  
  mapped_genes2 <- select(txdb, 
                          keys = annotated_df2$tx_name, 
                          keytype = "TXNAME", 
                          columns = c("GENEID"))
  
  annotated_df1 <- cbind(annotated_df1, mapped_genes1)
  annotated_df1 <- annotated_df1 |>
    dplyr::mutate(gene_id1 = GENEID) |>
    dplyr::select(-c(TXNAME, GENEID))
  
  annotated_df2 <- cbind(annotated_df2, mapped_genes2)
  annotated_df2 <- annotated_df2 |>
    dplyr::mutate(gene_id2 = GENEID) |>
    dplyr::select(-c(TXNAME, GENEID))
  
  annotated_df <- annotated_df1 |>
    dplyr::left_join(annotated_df2, by = "ix_id")
  
  final_df <- df |>
    dplyr::left_join(annotated_df, by = "ix_id") |>
    dplyr::select(-c(ix_id))
  
  return(final_df)
}
