#' annotateDIR
#'
#' @description
#' This function reads differentially-interacted regions (DIRs) and annotate with hg38 information
#' 
#' @author Tsunghan Hsieh
#'
#' @param input a character string specifying the name and path of the input file (.csv)
#' This file should be generated by getDIRWithNoReplicate()
#' @param output A character string specifying the name and path of output file
#' the default path is as same as the treatment file
#' the default output file name is the treatment file name prefixed with "annotated_"
#' @return a dataframe of the output data
#'
#' @export
#'
#' @importFrom here here
#' @importFrom assertthat assert_that
#' @importFrom tidyr separate#' 
#' @importFrom dplyr select filter left_join
#' @importFrom annotatr annotate_regions build_annotations
#' @importFrom GenomicRanges GRanges
#' @importFrom IRanges IRanges
#'
#' @examples
#' \dontrun{
#' df <- annotateDIR(
#' input = here::here("./Results/DIR/bg.csv"),
#' output = NULL
#' )
#' }
#' 
## TODO
## Run getDIRWithNoReplicate() to generate DIR files

annotateDIR <- function(input, output) {
  # check input ----------------------------------------------------------------
  
  assertthat::assert_that(is.character(input), 
                          msg = "The given argument is not a string.\n")
  
  assertthat::assert_that(grepl("\\.csv$", input), 
                          msg = "The given file to the input file is not a .csv file.\n")
  
  assertthat::assert_that(file.exists(here::here(input)), 
                          msg = "The given file path to the input file does not exist.\n")
  
  # check output ---------------------------------------------------------------
  
  if (is.null(output)) {
    # Use sub() to extract the desired part
    outputfname <- sub(".*/([^.]+)\\.csv", "\\1", input)
    outputfname <- paste0("annotated_", outputfname, ".csv")
    outputdir <- dirname(input)
    output <- here::here(outputdir, outputfname)
    remove(outputfname, outputdir)
  } else {
    assertthat::assert_that(is.character(output), 
                            msg = "The given argument is not a string.\n")
    
    assertthat::assert_that(grepl("\\.csv$", output), 
                            msg = "The given file to the output file is not a .csv file.\n")
    
    assertthat::assert_that(file.exists(here::here(dirname(output))), 
                            msg = "The given file path to the expertiment file does not exist.\n")
  }
  
  
  # Read DIRs ------------------------------------------------------------------
  input_df <- read.csv(input)
  input_df <- input_df[,2:ncol(input_df)]
  input_df <- input_df |>
    dplyr::mutate(id = paste0(region1, "-", region2)) |>
    dplyr::mutate(strand = "*") |>
    dplyr::select(id, chr, region1, region2, strand, logFC, logCPM, PValue)
  colnames(input_df) <- c("id", "chr", "start", "end", "strand", "logFC", "logCPM", "PValue")
  col_names <- c("chr", "start", "end", "strand")
  
  annotated_df <- .get_annotated_gr(input_df, col_names)
  
  ## Annotation containing multiple exons or introns are not required
  annotated_df <- annotated_df |>
    dplyr::distinct(id, .keep_all = TRUE) |>
    dplyr::select(-c(chr, start,end,strand))
  
  final_df <- input_df |>
    dplyr::left_join(annotated_df, by = "id")
  
  write.csv(final_df, output)
  message("Done! The output is written in: ", output)
  return(annotated_df)
}


#' .get_annotated_gr
#'
#' @description
#' This function annotate differentially-interacted regions (DIRs) with given genome information
#' 
#' @author Tsunghan Hsieh
#'
#' @param input a character string specifying the input dataframe
#' @param col_names a character vector specifying the column names of return dataframe
#' @return a dataframe of annotated file
#' @examples
#' \dontrun{
#' df <- .get_annotated_gr(
#' df = input_df,
#' col_names = col_names
#' )
#' }
.get_annotated_gr <- function(df, col_names) {
  
  # Check colnames -------------------------------------------------------------
  ## Filter out non-standard chr
  wanted <- paste0("chr", seq(1:23))
  wanted <- c(wanted, "chrX", "chrY")
  df <- df |>
    dplyr::filter(chr %in% wanted)
  
  # Create GRange objects ------------------------------------------------------
  ## For creating GRange objects
  ## The id column is used as the identity column
  
  gr <- GenomicRanges::GRanges(
    seqnames = df$chr,
    ranges = IRanges::IRanges(start = df$start, end = df$end, strand = df$strand),
    id = df$id
  )
  
  # Annotate -------------------------------------------------------------------
  ## Select annotations for intersection with regions
  ## Note inclusion of custom annotation, and use of shortcuts
  annots <- c('hg38_basicgenes')
  
  ## Build the annotations (a single GRanges object)
  annotations <- annotatr::build_annotations(genome = 'hg38', annotations = annots)
  
  df_annotated <- annotatr::annotate_regions(
    regions = gr,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = TRUE)
  
  df_annotated <- data.frame(df_annotated)
  
  df_annotated <- df_annotated |>
    dplyr::select(-c(strand))
  
  ## Add the strand information back
  df_strand <- df |>
    dplyr::select(id, strand)
  
  df_annotated <- df_annotated |>
    dplyr::left_join(df_strand, by = "id")
  
  # Generate the final data frame ----------------------------------------------
  final_data <- df_annotated |>
    dplyr::select(id, seqnames, start, end, strand)
  
  df_annotated <- df_annotated |>
    dplyr::select(-c(id, strand))
  
  colnames(final_data) <- c("id", col_names)
  final_data <- cbind(final_data, df_annotated[,5:ncol(df_annotated)])
  
  return(final_data)
}
